forest/cli/src/humantoken.rs

```diff
@@ -0,0 +1,653 @@
+// Copyright 2019-2023 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0, MIT
+//! This module defines a [parser](parse()) and
+//! [pretty-printer](TokenAmountPretty::pretty) for
+//! `TokenAmount`
+//!
+//! See the `si` module source for supported prefixes.
+
+pub use parse::parse;
+pub use print::TokenAmountPretty;
+
+/// SI prefix definitions
+mod si {
+    use bigdecimal::BigDecimal;
+
+    // Use a struct as a table row instead of an enum
+    // to make our code less macro-heavy
+    #[derive(Debug, Clone, Copy, PartialEq, Eq)]
+    pub struct Prefix {
+        /// `"micro"`
+        pub name: &'static str,
+        /// `[ "μ", "u" ]`
+        pub units: &'static [&'static str],
+        /// `-6`
+        pub exponent: i8,
+        /// `"0.000001"`
+        pub multiplier: &'static str,
+    }
+
+    impl Prefix {
+        // ENHANCE(aatifsyed): could memoize this if it's called in a hot loop
+        pub fn multiplier(&self) -> BigDecimal {
+            self.multiplier.parse().unwrap()
+        }
+    }
+
+    /// Biggest first
+    macro_rules! define_prefixes {
+        ($($name:ident $symbol:ident$(or $alt_symbol:ident)* $base_10:literal $decimal:literal),* $(,)?) =>
+            {
+                // Define constants
+                $(
+                    #[allow(non_upper_case_globals)]
+                    pub const $name: Prefix = Prefix {
+                        name: stringify!($name),
+                        units: &[stringify!($symbol) $(, stringify!($alt_symbol))* ],
+                        exponent: $base_10,
+                        multiplier: stringify!($decimal),
+                    };
+                )*
+
+                /// Biggest first
+                // Define top level array
+                pub const SUPPORTED_PREFIXES: &[Prefix] =
+                    &[
+                        $(
+                            $name
+                        ,)*
+                    ];
+            };
+    }
+
+    define_prefixes! {
+        quetta	Q	30	1000000000000000000000000000000,
+        ronna	R	27	1000000000000000000000000000,
+        yotta	Y	24	1000000000000000000000000,
+        zetta	Z	21	1000000000000000000000,
+        exa	    E	18	1000000000000000000,
```

Comment:
This doesn't align correctly when viewed on GitHub. Using spaces might make it more consistent.

Reply:
Github and (my) VSCode have different tab stops, so both can't look right. 
Favouring github here since that's what most readers will use

Reply:
Eh, still broken, I've just manually aligned with spaces